src/Pact/Native/Trans/Types.hs

```diff
@@ -0,0 +1,263 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE DeriveFunctor #-}
+{-# LANGUAGE DeriveFoldable #-}
+{-# LANGUAGE DeriveTraversable #-}
+{-# LANGUAGE ForeignFunctionInterface #-}
+
+-- |
+-- Module      :  Pact.Native.Trans.Types
+-- Copyright   :  (C) 2022 John Wiegley
+-- License     :  BSD-style (see the file LICENSE)
+-- Maintainer  :  John Wiegley <john@kadena.io>
+--
+-- Operators and math built-ins.
+--
+
+module Pact.Native.Trans.Types where
+
+import Control.Exception
+import Data.Decimal (Decimal)
+import Data.Int
+import Data.Word
+import Data.Ratio
+import Foreign.C.String (withCString, peekCString)
+import Foreign.C.Types
+import Foreign.Marshal.Alloc (alloca, free)
+import Foreign.Ptr
+import Foreign.Storable
+import System.IO.Unsafe (unsafePerformIO)
+
+data TransResult a
+  = TransNumber !a
+  | TransNaN
+  | TransInf
+  | TransNegInf
+  deriving (Show, Eq, Ord, Functor, Foldable, Traversable)
+
+data MPZ = MPZ {
+  _mpzAlloc :: {-# UNPACK #-} !Int32,
+  _mpzSize  :: {-# UNPACK #-} !Int32,
+  _mpzD     :: {-# UNPACK #-} !(Ptr Limb)
+}
+
+instance Storable MPZ where
+    sizeOf _ = (16)
```

Comment:
Consider introducing a constant for this and derive other constants from it?

Reply:
I'm not sure what that would look like... I'm just carrying over the constant sizing and offsets of the C declarations. I don't know if the "logic" behind those numbers is worth trying to capture in Haskell.